File: CountPairsWithSum.java
// Question 2 Solution in Java:

import java.util.HashMap;

public class CountPairsWithSum {

    public static int countPairsWithSum(int[] nums, int target) {
        HashMap<Integer, Integer> numMap = new HashMap<>();
        int count = 0;

        for (int num : nums) {
            int complement = target - num;
            if (numMap.getOrDefault(complement, 0) > 0) {
                count += numMap.get(complement);
            }
            numMap.put(num, numMap.getOrDefault(num, 0) + 1);
        }
        return count;
    }

    public static void main(String[] args) {
        int[] nums = {1, 5, 7, -1, 5};
        int target = 6;
        System.out.println("Count of pairs is " + countPairsWithSum(nums, target));
    }
}

File: CountPairsWithSum.cpp
// Question 2 Solution in C++:
// 
// Count the number of pairs in an array with a given sum using a hash map.

// Given an array of integers and a target sum, return the number of pairs of integers in the array that add up to the target sum.
  
#include <iostream>
#include <unordered_map>
#include <vector>

int countPairsWithSum(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> numMap;
    int count = 0;

    for (int num : nums) {
        int complement = target - num;
        if (numMap[complement] > 0) {
            count += numMap[complement];
        }
        numMap[num]++;
    }
    return count;
}

int main() {
    std::vector<int> nums = {1, 5, 7, -1, 5};
    int target = 6;
    std::cout << "Count of pairs is " << countPairsWithSum(nums, target) << std::endl;
    return 0;
}
